hoomd/data/typeparam.py

```diff
@@ -4,12 +4,87 @@
 
 
-class TypeParameter:
-    """Store parameters per type."""
-    __slots__ = ('name', 'type_kind', 'param_dict')
+class TypeParameter(MutableMapping):
+    """Implement a type based mutable mapping.
+
+    *Implements the `collections.abc.MutableMapping` interface (``__delitem__``
+    is disallowed).*
+
+    `TypeParameter` instances extend the base Python mapping interface with with
+    smart defaults, value/key validation/processing, and advanced indexing.
+
+    .. rubric:: Indexing
+
+    For getting and setting values, multiple index formats are supported. The
+    base case is either a string representing the appropriate type or a tuple of
+    such strings if multiple types are required per key. Extending this an
+    iterator of the final types will retreive or set all keys in the iterator.
+    Likewise, for each item in the final tuple (if the expected key type is a
+    tuple of multiple string types), an iterator can be used instead of a string
+    which will result in all permutations of such iterators in the tuple. Both
+    advanced intexing methods can be combined.
+
+    Note:
+        Ordering in tuples does not matter.
+
+    Below are some example indexing values for single and multiple key indexing.
+
+    .. code-block:: python
+
+        # "A", "B", "C"
+        ["A", "B", "C"]
+        # ("A", "B")
+        ("A", "B")
+        # ("A", "B") and ("B", "C")
+        [("A", "B"), ("B", "C")]
+        # ("A", "B"), ("A", "C"), and ("A", "D")
+        ("A", ["B", "C", "D"])
+
+
+    .. rubric:: Defaults and setting values
+
+    `TypeParameter` instances have default values that can be accessed via
+    ``default`` which will be used for all types not defined. In addition, when
+    the type parameter expects a `dict`-like object, the default will be updated
+    with the set value. This means that values that have defaults do not need to
+    be explicitly specified.
+
+    An example of "smart"-setting using the MD LJ potential,
+
+    .. code-block:: python
+
+        lj = hoomd.md.pair.LJ(nlist=hoomd.md.nlist.Cell())
+        # params is a TypeParameter object.
+        # We set epsilon to have a default but sigma is still required
+        lj.params.default = {"epsilon": 4}
+        print(lj.params.default)
+        # {"epsilon": 4.0, "sigma": hoomd.data.typeconverter.RequiredArg}
+        # We do no need to specify epsilon to use new default value when
+        # setting
+        lj.params[("A", "B")] = {"sigma": 1.0}
+        print(lj.params[("A", "B")])
+        # {"epsilon": 4.0, "sigma": 1.0}
+
+    Note:
+        Setting values for fictitious but valid types will not trigger an error
+        before calling `hoomd.Simulation.run`, but attempts to access that data
+        or set values for fictitious types afterwards will result in
+        ``KeyError`` exceptions.
+
+    Warning:
+        For nested data structures, editing the internal stuctures such as a
```

Comment:
Start off with something positive like "Values in this mapping are returned by copy, not reference. To modify nested data structures, such as the union shape intergrators in HPMC and the table pair potential in MD, read the current value, modify it, then write a new parameter (example)."

Also, don't users "anti-pattern". Most users won't know what that means.